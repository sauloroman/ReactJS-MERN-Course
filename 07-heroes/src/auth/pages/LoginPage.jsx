import { useContext } from "react"
import { useNavigate } from "react-router-dom"
import { AuthContext } from "../context"

export const LoginPage = () => {

  const { login } = useContext( AuthContext )
  const navigate = useNavigate()

  const onLogin = () => {

    const lastLogin = localStorage.getItem('lastPath') || '/'

    login('Roman Nava')

    navigate( lastLogin, { replace: true })
  }

  return (
    <div className="login">
      <div className="login__container">
        <h1 className="login__title">Ingresar</h1>

        <button onClick={ onLogin } className="login__button">Login</button>
      </div>
    </div>
  )
}
