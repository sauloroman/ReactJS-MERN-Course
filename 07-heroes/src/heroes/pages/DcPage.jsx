import { HeroList } from "../components"

export const DcPage = () => {
  return (
    <>
      <h1 className="title__page">Dc Comics</h1>
      <HeroList publisher={'DC Comics'} />
    </>
  )
}
