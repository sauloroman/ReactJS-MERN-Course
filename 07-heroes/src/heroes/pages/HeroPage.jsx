import { useMemo } from 'react'
import { useNavigate, useParams, Navigate } from 'react-router-dom'
import { getHeroById } from '../helpers'

export const HeroPage = () => {

  const { id } = useParams()
  const navigate = useNavigate()

  const hero = useMemo(() => getHeroById(id), [id])

  const onNavigateBack = () => {
    navigate(-1)
  }

  if ( !hero ) return <Navigate to='/marvel' />

  return (
    <div className='hero'>
      <div className="hero__pic">
        <img 
          src={`/assets/heroes/${id}.jpg`} 
          alt={hero.superhero} 
          className="hero__img" 
        />
      </div>
      <div className="hero__info">
        <h2 className="hero__page">{hero.superhero}</h2>

        <ul className="hero__list">
          <li className="hero__item">
            Alter ego: <span className="hero__data">{hero.alter_ego}</span>
          </li>
          <li className="hero__item">
            Personajes: <span className='hero__data'>{ hero.characters }</span>
          </li>
          <li className="hero__item">
            Primera ApariciÃ³n: <span className='hero__data'>{ hero.first_appearance }</span>
          </li>
          <li className="hero__item">
            ID: <span className='hero__data'>{ hero.id }</span>
          </li>
          <li className="hero__item">
            Publicador: <span className='hero__data'>{ hero.publisher }</span>
          </li>
        </ul>

        <button onClick={ onNavigateBack } className="hero__button">Regresar</button>
      </div>
    </div>
  )
}
