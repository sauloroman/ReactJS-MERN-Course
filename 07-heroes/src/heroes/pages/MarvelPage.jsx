import { HeroList } from "../components"

export const MarvelPage = () => {
  return (
    <>
      <h1 className="title__page">Marvel Comics</h1>
      <HeroList publisher={'Marvel Comics'} />
    </>
  )
}
