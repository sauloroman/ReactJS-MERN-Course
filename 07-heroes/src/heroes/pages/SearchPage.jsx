import queryString from 'query-string'
import { useLocation, useNavigate } from "react-router-dom"
import { getHeroesByName } from '../helpers'
import { useForm } from '../../../src/hooks/useForm'
import { HeroCard } from '../components/HeroCard'

export const SearchPage = () => {

  const navigate = useNavigate()
  const location = useLocation()

  const { q = '' } = queryString.parse( location.search )
  const heroes = getHeroesByName( q )

  const showSearch = ( q.length === 0 )
  const showError = ( q.length > 0 ) && heroes.length === 0

  const { searchHero, onInputChange } = useForm({
    searchHero: q
  })

  const onSearchSubmit = (e) => {
    e.preventDefault()
    navigate(`?q=${searchHero}`)
  }
  
  return (
    <div className="search">
      <h1 className="search__title">Búsqueda</h1>

      <div className="search__grid">
        <div className="search__content">
          <h2 className="search__subtitle">Encuentra tu heroe</h2>
          <form aria-label='form' onSubmit={ onSearchSubmit } className="search__form">
            <input
              placeholder="Ingresa el nombre de tu heroe favorito" 
              type="text" 
              className="search__input"
              autoComplete="off"
              name='searchHero'
              value={ searchHero }
              onChange={ onInputChange }
            />
            <button className="search__button">Buscar</button>
          </form>
        </div>
        <div className="search__alerts">
          <h2 className="search__subtitle">Resultados</h2>

          <div className={`search__alert search__alert--success ${ showSearch || 'hidden' }`}>
            Busca un heroe
          </div>

          <div aria-label='alert-error' className={`search__alert search__alert--error ${ showError || 'hidden' }`}>
            No se encontró al heroe <span className='strong'>{q}</span>
          </div>

          <div className="search__results">
            {
              heroes.map( hero => (
                <HeroCard key={hero.id} {...hero} />
              ))
            }
          </div>

        </div>

      </div>

    </div>
  )
}
