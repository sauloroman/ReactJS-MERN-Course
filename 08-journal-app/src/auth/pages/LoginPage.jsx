import { Link as RouterLink } from 'react-router-dom'
import { Button, Grid2, TextField, Typography, Link, Alert } from "@mui/material"
import { Google } from '@mui/icons-material'
import { AuthLayout } from "../layout/AuthLayout"
import { useForm } from '../../hooks'
import { useDispatch, useSelector } from 'react-redux'
import { startCreatingUserWithEmailAndPasswordThunk, startGoogleSignInThunk, startLoginWithEmailAndPasswordThunk } from '../../store/auth'

const formData = {
  email: '',
  password: ''
}

export const LoginPage = () => {

  const dispatch = useDispatch()

  const { status, errorMessage } = useSelector( state => state.auth )
  const { formState, email, password, onInputChange } = useForm( formData )

  const onSubmitLogin = (e) => {
    e.preventDefault()
    dispatch( startLoginWithEmailAndPasswordThunk(formState) )
  }

  const onGoogleSignIn = () => {
    dispatch( startGoogleSignInThunk() )
  }

  return (
    <AuthLayout title="Iniciar Sesi칩n">
      <form 
        onSubmit={ onSubmitLogin }
        aria-label="form-login"
        action="#"
        className="animate__animated animate__fadeIn animate__faster"
      >

        <Grid2 direction='column' spacing={2} container>
          
          <Grid2 item>
            <TextField 
              label="Correo electr칩nico"
              type="email"
              placeholder="correo@gmail.com"
              fullWidth
              name="email"
              value={email}
              onChange={onInputChange}
            />
          </Grid2>

          <Grid2 item>
            <TextField 
              label="Contrase침a"
              type="password"
              placeholder="Ingresa su contrase침a"
              fullWidth
              name="password"
              value={password}
              onChange={onInputChange}
              slotProps={{'data-testid': 'input-password'}}
            />
          </Grid2>

          <Grid2
            display={ !!errorMessage ? '' : 'none' } 
            container 
            sx={{ mb: 2, mt: 1 }} spacing={2}
          >
            <Grid2 item size={{ xs: 12, sm: 6 }}>
              <Alert severity='error'>{errorMessage}</Alert>
            </Grid2>
          </Grid2>

          <Grid2 container spacing={2} sx={{ mb: 2, mt: 1 }}>
            <Grid2 item size={{ xs: 12, sm: 6 }}>
              <Button
                sx={{ padding: 1 }}
                type="submit"
                variant="contained"
                fullWidth
              >
                <Typography fontSize={14}>Ingresar</Typography>
              </Button>
            </Grid2>
            <Grid2 item size={{ xs: 12, sm: 6 }}>
              <Button
                onClick={ onGoogleSignIn }
                sx={{ padding: 1 }}
                variant="outlined"
                fullWidth
                aria-label="google-btn"
              >
                <Google sx={{ height: 20 }} />
                <Typography fontSize={14} sx={{ ml: 1 }}>Google</Typography>
              </Button>
            </Grid2>
          </Grid2>

          <Grid2 container direction='row' justifyContent='end' >
            <Link component={RouterLink} color="inherit" to="/auth/register">Crear una cuenta</Link>
          </Grid2>

        </Grid2>

      </form>
    </AuthLayout>
  )
}
