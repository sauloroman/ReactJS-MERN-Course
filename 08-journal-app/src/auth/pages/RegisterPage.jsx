import { Link as RouterLink } from 'react-router-dom'
import { Alert, Button, Grid2, Link, TextField, Typography } from "@mui/material"
import { AuthLayout } from "../layout/AuthLayout"
import { useForm } from '../../hooks'
import { useMemo, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startCreatingUserWithEmailAndPasswordThunk } from '../../store/auth'

const formData = {
  email: '',
  password: '',
  displayName: '',
}

const formValidations = {
  email: [ (value) => value.includes('@'), 'El correo debe tener una @'],
  password: [ (value) => value.length >= 6, 'El password debe de tener más de 6 letras'],
  displayName: [(value) => value.length >= 1, 'El nombre es obligatorio']
}

export const RegisterPage = () => {

  const dispath = useDispatch()
  const [formSubmitted, setFormSubmitted] = useState(false)
  const { status, errorMessage } = useSelector( state => state.auth )
  const isCheckinAuthentication = useMemo(() => status === 'checking', [status])

  const {
    formState, displayName, email, password, onInputChange,
    isFormValid, displayNameValid, emailValid, passwordValid,
  } = useForm( formData, formValidations )

  const onSubmitRegister = (e) => {
    e.preventDefault()
    setFormSubmitted( true )

    if ( !isFormValid ) return;

    dispath( startCreatingUserWithEmailAndPasswordThunk(formState) )
  }

  return (
    <AuthLayout title="Crear cuenta">
      <form
        onSubmit={ onSubmitRegister }
        className="animate__animated animate__fadeIn animate__faster" 
        action="#"
      >
        <Grid2 container direction={'column'} spacing={2}>

          <Grid2 item>
            <TextField 
              label="Nombre completo"
              type="text"
              placeholder="Ingrese su nombre completo"
              fullWidth
              name='displayName'
              value={displayName}
              onChange={onInputChange}
              error={ !!displayNameValid && formSubmitted }
              helperText={ displayNameValid }
            />
          </Grid2>

          <Grid2 item>
            <TextField 
              label="Correo"
              type="email"
              placeholder="correo@google.com"
              fullWidth
              name='email'
              value={email}
              onChange={ onInputChange }
              error={ !!emailValid && formSubmitted }
              helperText={emailValid}
            />
          </Grid2>

          <Grid2 item>
            <TextField 
              label="Contraseña"
              type="password"
              placeholder="Cree una contraseña"
              fullWidth
              name='password'
              value={password}
              onChange={onInputChange}
              error={ !!passwordValid && formSubmitted }
              helperText={ passwordValid }
            />
          </Grid2>

          <Grid2 container spacing={2} sx={{ mb: 2, mt: 1 }}>

            <Grid2 item size={{ xs: 12 }} display={ !!errorMessage ? '' : 'none' }>
              <Alert severity='error'>{errorMessage}</Alert>
            </Grid2>            

            <Grid2 item size={{ xs: 12 }}>
              <Button
                disabled={isCheckinAuthentication}
                type="submit"
                variant="contained"
                fullWidth
              >
                <Typography fontSize={14}>Crear cuenta</Typography>
              </Button>
            </Grid2>
          </Grid2>

          <Grid2 container direction={'row'} spacing={1} justifyContent={'start'}>
            <Typography>Ya tienes una cuenta?</Typography>
            <Link to={'/auth/login'} color='inherit' component={RouterLink}>Ingresar</Link>
          </Grid2>

        </Grid2>
      </form>
    </AuthLayout>
  )
}
