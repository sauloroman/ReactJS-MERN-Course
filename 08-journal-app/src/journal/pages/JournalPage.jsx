import { IconButton } from "@mui/material"
import { JournalLayout } from "../layout/JournalLayout"
import { AddOutlined } from "@mui/icons-material"
import { useDispatch, useSelector } from "react-redux"
import { NoteView, NothingSelectedView } from "../views"
import { startNewNoteThunk } from "../../store/journal/journal.thunk"

export const JournalPage = () => {

  const dispatch = useDispatch()
  const { isSaving, active } = useSelector( state => state.journal )

  const onClickNewNote = () => {
    dispatch( startNewNoteThunk() )
  }

  return (
    <JournalLayout>
      
      {
        !!active
        ? (<NoteView />)
        : (<NothingSelectedView />)
      }

      <IconButton
        onClick={ onClickNewNote }
        disabled={ isSaving }
        size="medium"
        sx={{
          color: "#fff",
          backgroundColor: 'error.main',
          ':hover': { backgroundColor: 'error.main', opacity: 0.9 },
          position: 'fixed',
          right: 50,
          bottom: 50
        }}
      >
        <AddOutlined sx={{ fontSize: 30 }}/>
      </IconButton>
    </JournalLayout>
  )
}
