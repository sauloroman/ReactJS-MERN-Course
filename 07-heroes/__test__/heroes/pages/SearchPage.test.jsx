import { fireEvent, render, screen } from '@testing-library/react'
import { SearchPage } from '../../../src/heroes/pages/SearchPage'
import { MemoryRouter } from 'react-router-dom'

const mockedUseNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate
}))

describe('Tests in SearchPage Component', () => {

  beforeEach(() => jest.clearAllMocks() )

  test('it must match with snapshot', () => {

    const {container} = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    )

    expect( container ).toMatchSnapshot()

  })

  test('it should show Batman hero and the input with the value from queryString', () => {

    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchPage />
      </MemoryRouter>
    )

    const input = screen.getByRole('textbox')
    expect( input.value ).toBe('batman')

    const batmanImg = screen.getByRole('img')
    expect( batmanImg.src ).toContain('/assets/heroes/dc-batman.jpg')

    const alertError = screen.getByLabelText('alert-error')
    expect( alertError.className ).toContain('hidden')

  })

  test('it should show an error if the hero is not found', () => {

    render(
      <MemoryRouter initialEntries={['/search?q=noexiste']}>
        <SearchPage />
      </MemoryRouter>
    )

    const alertError = screen.getByLabelText('alert-error')
    expect( alertError.className ).not.toContain('hidden')

  })

  test('it should navigate to the proper search page', () => {

    const searchHero = 'superman'

    render(
      <MemoryRouter initialEntries={['/search']}>
        <SearchPage />
      </MemoryRouter>
    )

    const inputBox = screen.getByRole('textbox')
    fireEvent.change( inputBox, { target: {name: 'searchHero', value: searchHero } } )

    const form = screen.getByLabelText('form')
    fireEvent.submit( form )

    expect( mockedUseNavigate ).toHaveBeenCalledWith(`?q=${searchHero}`)

  })

})