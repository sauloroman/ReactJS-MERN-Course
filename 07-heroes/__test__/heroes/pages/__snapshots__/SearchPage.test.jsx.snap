// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in SearchPage Component it must match with snapshot 1`] = `
<div>
  <div
    class="search"
  >
    <h1
      class="search__title"
    >
      Búsqueda
    </h1>
    <div
      class="search__grid"
    >
      <div
        class="search__content"
      >
        <h2
          class="search__subtitle"
        >
          Encuentra tu heroe
        </h2>
        <form
          aria-label="form"
          class="search__form"
        >
          <input
            autocomplete="off"
            class="search__input"
            name="searchHero"
            placeholder="Ingresa el nombre de tu heroe favorito"
            type="text"
            value=""
          />
          <button
            class="search__button"
          >
            Buscar
          </button>
        </form>
      </div>
      <div
        class="search__alerts"
      >
        <h2
          class="search__subtitle"
        >
          Resultados
        </h2>
        <div
          class="search__alert search__alert--success true"
        >
          Busca un heroe
        </div>
        <div
          aria-label="alert-error"
          class="search__alert search__alert--error hidden"
        >
          No se encontró al heroe 
          <span
            class="strong"
          />
        </div>
        <div
          class="search__results"
        />
      </div>
    </div>
  </div>
</div>
`;
